### PR TITLE
feat(mixin): add manual changes for pubsub and generate mixin code

### PR DESCRIPTION
**Pull request by cuiy0006**
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14756)

### CODE REVIEW

File: generator/internal/make_generators.cc [hunk 1]
```diff
@@ -61,7 +61,9 @@ std::vector<std::unique_ptr<GeneratorInterface>> MakeGenerators(
     YAML::Node const& service_config,
     std::vector<std::pair<std::string, std::string>> const& vars) {
   std::vector<MixinMethod> mixin_methods;
-  //  mixin_methods = GetMixinMethods(service_config, *service);
+  if (service->file()->name() == "google/pubsub/v1/pubsub.proto") {
```
Comment (dbolduc) [inline generator/internal/make_generators.cc:64]: I really like the tactic to introduce the new code in a single library first.

File: google/cloud/pubsub/BUILD.bazel [hunk 2]
```diff
@@ -38,6 +38,8 @@ cc_library(
     deps = [
         "//:common",
         "//:grpc_utils",
+        "@com_google_googleapis//google/cloud/location:location_cc_grpc",
```
Comment (dbolduc) [inline google/cloud/pubsub/BUILD.bazel:41]: necessary? It does not look like pubsub is mixing in the locations service.

File: google/cloud/pubsub/samples/samples.cc [hunk 3]
```diff
@@ -1034,8 +1034,8 @@ void OptimisticSubscribe(std::vector<std::string> const& argv) {
   namespace pubsub_admin = ::google::cloud::pubsub_admin;
   namespace pubsub = ::google::cloud::pubsub;
   namespace gc = ::google::cloud;
-  [](std::string project_id, std::string topic_id,
-     std::string subscription_id) {
+  [](std::string const& project_id, std::string topic_id,
+     std::string const& subscription_id) {
```
Comment (dbolduc) [inline google/cloud/pubsub/samples/samples.cc:1038]: FYI: I am racing with you in #14758. I don't think there will be any merge conflicts though. I think GitHub should be smart enough to handle it.

File: google/cloud/pubsub/BUILD.bazel (hunk 2 above)
Comment (dbolduc) [reply google/cloud/pubsub/BUILD.bazel:41]: Also how do you plan to update these `BUILD` files for other services? Manually?

In CMake, the equivalent of `//:grpc_utils` is depending on the mixin services. I am not sure Bazel's `layering_check` will allow that though.

**Review by dbolduc — APPROVED** [review]
This all looks good. What is the testing plan though?

## Unit testing

I think we want golden unit tests, at least in the stub that verifies that the mixin `StubInterface` is used, not the service's.

It looks like there is a way to define our own `*.yaml` for the golden files.

https://github.com/googleapis/google-cloud-cpp/blob/a695357095b06dc11826df10394d50733d0d98d8/generator/integration_tests/golden_config.textproto#L72

Then we would add a unit test here:

https://github.com/googleapis/google-cloud-cpp/tree/main/generator/integration_tests/tests

## Integration Testing

I think we also need an integration test that proves that these APIs work against production. We typically use spanner's admin APIs for integration testing the generator.

It looks like we can eventually have something for the LRO mixin.

https://github.com/googleapis/googleapis/blob/78d1fb208e4af3022a0aaf27bafa578ff326326e/google/spanner/admin/database/v1/spanner.yaml#L7

---

Also, this API is now obsolete:

https://github.com/googleapis/google-cloud-cpp/blob/a695357095b06dc11826df10394d50733d0d98d8/google/cloud/pubsub/options.h#L415-L429

We should deprecate it, and recommend that applications use the new mixed-in APIs. Can you open an issue?

File: google/cloud/pubsub/BUILD.bazel (hunk 2 above)
Comment (cuiy0006) [reply google/cloud/pubsub/BUILD.bazel:41]: I should remove `location_cc_grpc`. I planned to add the dependencies for each library. Do you know where I can add the mixin dependencies for bazel like what I did for cmake in `google/cloud/google_cloud_cpp_grpc_utils.cmake`? That will save me a lot of efforts.

Comment (dbolduc) [reply google/cloud/pubsub/BUILD.bazel]: https://github.com/googleapis/google-cloud-cpp/blob/3601f912caad97b242059ac1af336becc71d659e/google/cloud/BUILD.bazel#L166-L176

Hopefully bazel lets us pick up these dependencies transitively... :crossed_fingers:

Comment (cuiy0006) [reply google/cloud/pubsub/BUILD.bazel]: Great! It works.

**Comment by cuiy0006** [comment]
I created an issue for adding tests for mixin.

https://github.com/googleapis/google-cloud-cpp/issues/14762

**Pull request merged**